Add devcontainer to improve developer experience for contributors (#3942)
dapr/docs: README.md

The [daprdocs](./daprdocs) directory contains the hugo project, markdown files, and theme configurations.

## Setup with a devcontainer

> This repository comes with a [devcontainer](/.devcontainer/devcontainer.json) configuration that will automatically install all the required dependencies and tools to build and run the docs.

Reviewer: ```suggestion
This repository comes with a [devcontainer](/.devcontainer/devcontainer.json) configuration that automatically installs all the required dependencies and tools to build and run the docs.
```

This devcontainer can be used to develop locally with VSCode or via GitHub Codespaces completely in the browser.

### Pre-requisites

- [Docker Desktop](https://www.docker.com/products/docker-desktop)
- [VSCode](https://code.visualstudio.com/download)

### Environment setup

1. [Fork](https://github.com/dapr/docs/fork) and clone this repository.

2. Open the forked repository in VS Code

```sh
code .
```

> 3. When prompted, click "Reopen in Container" to open the repository in the devcontainer.

Reviewer 2: I tried your PR out with Goland and didn't get prompted. For VS Code I did get the prompt. It might be worth documenting the steps a user should take if they don't get prompted by their IDE. 

It also took quite a while to load and get to this point, specifically, it took: 3.3588 minutes for it to start my devcontainer for me to be able to proceed:
```
[201528 ms] Start: Run in container: cd '/vscode/vscode-server/extensionsCache' && ls -t | tail -n +50 | xargs rm -f
[201532 ms] 
[201532 ms] 
[238758 ms] Port forwarding connection from 49714 > 41037 > 41037 in the container.
[238758 ms] Start: Run in container: /home/vscode/.vscode-server/bin/0ee08df0cf4527e40edc9aa28f4b5bd38bbff2b2/node -e 
[238893 ms] Port forwarding 49714 > 41037 > 41037 stderr: Connection established
[243903 ms] Port forwarding 49714 > 41037 > 41037 stderr: Remote close
[243917 ms] Port forwarding 49714 > 41037 > 41037 terminated with code 0 and signal null.
[376402 ms] Port forwarding 49714 > 41037 > 41037: Local close
```

Reviewer: Was this container running locally or on GitHub?

Reviewer 2: I followed the instructions provided in the readme- I ran this locally. I wan't sure if @marcduiker's also took a while? It's just something I wanted to note since it does take a bit more time now.

Author: @cicoyle When I try it with VSCode it takes about 26 seconds (I've deleted the container and the image first). When I close and reopen VSCode with the container it takes about 8 seconds.

The devcontainer support for JetBrains products is still very new (was announced Dec 2023) so perhaps they need to improve their tooling a bit?

Author: The README does mention VSCode as a prerequisite (or use GitHub CodeSpaces).  I'll add a sentence that other IDEs (that support devcontainers) could be used, but might not result in the best experience.

Note that the devcontainer.sjon has some VSCode specific configuration, such as loading extensions which are useful for editing docs (Spellchecker, markdown linting, hugo support).

> 2. Open the forked repository in VS Code

Reviewer 3: ```suggestion
1. Open the forked repository in VS Code
```

This is a nit, but Markdown automatically numbers lists with the correct order - keeping all items as `1.` helps when we edit later, so we don't have to remember to renumber lists if we add or remove an item.

> 3. When prompted, click "Reopen in Container" to open the repository in the devcontainer.

Reviewer 3: ```suggestion
1. When prompted, click "Reopen in Container" to open the repository in the devcontainer.
```

Continue with the [Run local server](#run-local-server) steps.

## Setup without a devcontainer

### Pre-requisites

- [Hugo extended version](https://gohugo.io/getting-started/installing)
- [Node.js](https://nodejs.org/en/)

### Environment setup

1. Ensure pre-requisites are installed.

> 2. [Fork](https://github.com/dapr/docs/fork) and clone this repository.

Reviewer 3: ```suggestion
1. [Fork](https://github.com/dapr/docs/fork) and clone this repository.
```

> 3. Change to daprdocs directory:

Reviewer 3: ```suggestion
1. Change to daprdocs directory:
```

Reviewer 3: can apply to rest of numbered lists...

3. Navigate to `http://localhost:1313/`

## Update docs

1. Ensure you are in your forked repo

> 1. Create new branch

Reviewer 2: While you are updating this, can you update the bulleted numbers. It looks as though each step is `#1`

Author: Done
